src/compat/tracker.jl

```diff
@@ -12,6 +12,8 @@ using .Tracker: Tracker,
 using Compat: eachcol
 using LinearAlgebra
 
+# Broadcasting here breaks Tracker for some reason
+maporbroadcast(f, x::Union{AbstractArray, TrackedArray, AbstractArray{<:TrackedReal}}...) = map(f, x...)
 maporbroadcast(f, x::TrackedArray...) = f.(x...)
```

Comment:
The definitions of `maporbroadcast` are quite dangerous since they easily lead to invalidations, similar to
```julia
julia> f(x::AbstractArray...) = 1
f (generic function with 1 method)

julia> f(x::Matrix...) = 2
f (generic function with 2 methods)

julia> f()
2
```
Fortunately, it seems to work in the case of `Union`s (at least with Julia 1.5.2),
```julia
julia> g(x::AbstractVector...) = 1
g (generic function with 1 method)

julia> g(x::AbstractVecOrMat...) = 2
g (generic function with 2 methods)

julia> g()
1
```
so this definition here might be the most problematic one.

Reply:
I mean we never really call the function without arguments. And unless there is a better solution that works, invalidations are a secondary concern.

Reply:
I think it would be easy to fix this issue - and at the same time improve the current design and implementation of `maporbroadcast`. Hence I don't think this problem should be a secondary concern.

Instead of implementing `maporbroadcast(f, x...)` for special types of `x` one could just define two traits `MapStyle` and `BroadcastStyle` (well, maybe use a different name...), and then define
```julia
maporbroadcast(f, x::AbstractArray...) = maporbroadcast(f, maporbroadcaststyle(x...), x...)

struct MapStyle end
struct BroadcastStyle end

maporbroadcast(f, ::MapStyle, x::AbstractArray...) = map(f, x...)
maporbroadcast(f, ::BroadcastStyle, x::AbstractArray...) = f.(x...)

maporbroadcaststyle(::Tuple{}) = MapStyle() # or BroadcastStyle, but maybe the type inference of map is better?
maporbroadcaststyle(::AbstractArray) = BroadcastStyle()
function maporbroadcaststyle(x::AbstractArray, y::AbstractArray, ys::AbstractArray...)
    return maporbroadcaststyle(maporbroadcaststyle(x), maporbroadcaststyle(y, ys...))
end

# promotion rule
maporbroadcaststyle(::MapStyle, ::MapStyle) = MapStyle()
maporbroadcaststyle(::BroadcastStyle, ::BroadcastStyle)
maporbroadcaststyle(::MapStyle, ::BroadcastStyle) = MapStyle()
maporbroadcaststyle(::BroadcastStyle, ::MapStyle) = MapStyle()

# to be implemented for TrackedArray etc.
maporbroadcaststyle(::TrackedArray) = MapStyle()
```

BTW I wanted to have a look at the corresponding implementation in DistributionsAD but I noticed that actually `maporbroadcast` is removed there.

Reply:
Please open another PR with this suggestion after this is merged.

Reply:
Your implementation is adding significant complexity to the code so it would be nice to evaluate its benefits in a separate thread, e.g. how many invalidations are eliminated. This PR is just a bugfix.

Reply:
IMO the code is not complex, it's just what you actually want to achieve (base whether `map` or `broadcast` is used on the arguments) without listing all possible combinations of arrays. Thus most importantly, it is easily extendable by just defining `maporbroadcaststyle` for arrays that should use broadcasting (or not, whatever the default is).